src/coreComponents/common/KernelLaunchSelectors.hpp

```diff
@@ -0,0 +1,167 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file KernelLaunchSelectors.hpp
+ */
+
+#ifndef GEOS_COMMON_KERNELLAUNCHSELECTORS_HPP
+#define GEOS_COMMON_KERNELLAUNCHSELECTORS_HPP
+
+namespace geos
+{
+namespace internal
+{
+
+template< typename T, typename LAMBDA >
+void kernelLaunchSelectorThermalSwitch( T value, LAMBDA && lambda )
+{
+  static_assert( std::is_integral< T >::value, "kernelLaunchSelectorThermalSwitch: type should be integral" );
+
+  switch( value )
+  {
+    case 0:
+    { lambda( std::integral_constant< T, 0 >() ); return; }
+    case 1:
+    { lambda( std::integral_constant< T, 1 >() ); return; }
+
+    default:
+    { GEOS_ERROR( "Unsupported thermal state: " << value ); }
+  }
+}
+
+template< typename T, typename LAMBDA >
+void kernelLaunchSelectorCompThermSwitch( T value, bool const isThermal, LAMBDA && lambda )
+{
+  static_assert( std::is_integral< T >::value, "kernelLaunchSelectorCompSwitch: value type should be integral" );
+
+  if( isThermal )
```

Comment:
instead of all these nested switches I think that you can do:

````c++
auto invokeLambda = [&] (auto val1, bool const isThermal) {
    if (isThermal) {
        lambda(val1, std::integral_constant<int, 1>{});
    } else {
        lambda(val1, std::integral_constant<int, 0>{});
    }
};

switch(value)
{
    case 1:
        invokeLambda(std::integral_constant<T, 1>{}, isThermal);  
        return;
    case 2:
        invokeLambda(std::integral_constant<T, 2>{}, isThermal); 
        return;
    default:
        return;
}
````


}

Reply:
Good suggestion... added something similar

template< typename S, typename T, typename LAMBDA >
void invokePhaseDispatchLambda ( S val, T numPhases, LAMBDA && lambda )
{
  if( numPhases == 1 )
  {
    lambda( val, std::integral_constant< T, 1 >());
    return;
  }
  else if( numPhases == 2 )
  {
    lambda( val, std::integral_constant< T, 2 >());
    return;
  }
  else if( numPhases == 3 )
  {
    lambda( val, std::integral_constant< T, 3 >());
    return;
  }
  else
  {
    GEOS_ERROR( "Unsupported state: " << numPhases );
  }
};